### DISCUSSION TITLE
How to close popup in PopupViewModel

### DISCUSSION THREAD

**Discussion by JaehwanMango** [CommunityToolkit/Maui#691] [Q&A] [answered]
[Docs = popup](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/views/popup)

```
// Code Behind
void OnYesButtonClicked(object? sender, EventArgs e) => Close(true);
void OnNoButtonClicked(object? sender, EventArgs e) => Close(false);
```
`toolkit:Popup` have function the close
I can't close in PopupViewModel

**Comment by pictos** [comment, marked as answer]
@JaehwanMango we don't provide a Command for that, if you want to do it on your VM, you can use our [EventToCommandBehavior](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/behaviors/event-to-command-behavior)

**Reply by JaehwanMango** [reply]
Hi, @pictos 
Thank you for reply

![image](https://user-images.githubusercontent.com/107899309/195746249-c98190f9-9234-44d2-b238-5cee06e905f9.png)

Popup.Close() is not event
how to use behavior?

**Reply by pictos** [reply]
You're right, sorry.

So there're a couple of ways of doing that...
1. You can use MessaingCenter or any other event-based pattern;
2. You can create on your VM a  `Command<IPopup>` and call the `OnClosed` method;
3. You can use the code-behind, this doesn't break the pattern, since controlling the visual state of the View isn't a VM responsibility (but you can disagree with this one if you wish);
4. You can use triggers.

Here's an idea behind option 2.

```xaml
<mct:Popup
	x:Name="this">
	<Frame
		BackgroundColor="Red"
		HeightRequest="50"
		WidthRequest="50">
		<Frame.GestureRecognizers>
			<TapGestureRecognizer Command="{Binding CloseCommand}" CommandParameter="{Binding Source={x:Reference this}}" />
		</Frame.GestureRecognizers>
	</Frame>
</mct:Popup>
```

```csharp
class VM
{
	public Command<IPopup> CloseCommand { get; }


	public VM()
	{
		CloseCommand = new Command<IPopup>(CloseCommandExecute);
	}

	void CloseCommandExecute(IPopup popup)
	{
		popup.OnClosed();
	}

}
```

**Comment by CliffAgius** [comment]
Closed as answered
